pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,258 @@
+# limitations under the License.
+"""
+This module contains functions for computing the finite-difference gradient
+of a quantum tape.
+"""
+# pylint: disable=protected-access,too-many-arguments
+import numpy as np
+from scipy.special import factorial
+
+import pennylane as qml
+
+
+def finite_diff_stencil(n, order, form):
+    r"""Generate the finite difference stencil (shift and coefficients)
+    for various derivatives, accuracy, and form.
+
+    Args:
+        n (int): Positive integer specifying the derivative. ``n=1``
+            corresponds to the first derivative.
+        order (int): Positive integer referring to the approximation order of the
+            returned stencil. E.g., ``order=1`` corresponds to a stencil
+            that returns a first-order approximation to the derivative.
+        form (str): one of ``"forward"``, ``"center"``, or ``"backward"``.
+
+    Returns:
+        array[float]: a ``(2, N)`` array. The first row corresponds to the
+        coefficients, and the second row corresponds to the shifts.
+
+    **Example**
+
+    >>> finite_diff_stencil(n=1, order=1, form="forward")
+    array([[-1.,  1.],
+           [ 0.,  1.]])
+
+    The first row corresponds to the coefficients, and the second corresponds
+    to the shifts. For example, this results in the linear combination:
+
+    .. math:: \frac{-y(x_0) + y(x_0 + h)}{h}
+
+    where :math:`h` is the finite-difference step-size.
+
+    More examples of finite-difference stencils:
+
+    >>> finite_diff_stencil(n=1, order=2, form="center")
+    array([[-0.5,  0.5],
+           [-1. ,  1. ]])
+    >>> finite_diff_stencil(n=2, order=2, form="center")
+    array([[-2.,  1.,  1.],
+           [ 0., -1.,  1.]])
+    """
+    if n < 1 or not isinstance(n, int):
+        raise ValueError("Derivative order n must be a positive integer.")
+
+    if order < 1 or not isinstance(order, int):
+        raise ValueError("Approximation order must be a positive integer.")
+
+    num_points = order + 2 * np.floor((n + 1) / 2) - 1
+    N = num_points + 1 if n % 2 == 0 else num_points
+
+    if form == "forward":
+        shifts = np.arange(N, dtype=np.float64)
+
+    elif form == "backward":
+        shifts = np.arange(-N + 1, 1, dtype=np.float64)
+
+    elif form == "center":
+        if order % 2 != 0:
+            raise ValueError("Centered finite-difference requires an even order.")
+
+        N = num_points // 2
+        shifts = np.arange(-N, N + 1, dtype=np.float64)
+
+    else:
+        raise ValueError(f"Unknown form {form}. Must be one of 'forward', 'backward', 'center'.")
+
+    A = shifts ** np.arange(len(shifts)).reshape(-1, 1)
+    b = np.zeros_like(shifts)
+    b[n] = factorial(n)
+    coeffs = np.linalg.solve(A, b)
+
+    stencil = np.stack([coeffs, shifts])
+    stencil[0, np.abs(stencil[0, :]) < 1e-10] = 0
+    stencil = stencil[:, ~np.all(stencil == 0, axis=0)]
+    stencil = stencil[:, np.argsort(np.abs(stencil)[1])]
+    return stencil

```

Comment:
Sorry, I am totally lost about what is going on :) Is this some kind of optimiser that finds the best shifts for finite diff? Would it then be better called `get_best_shifts` or so?

Reply:
Sorry, I should have commented this better. I've just updated it:

```
coeffs_and_shifts = np.stack([coeffs, shifts])

# remove all small coefficients and shifts
coeffs_and_shifts[np.abs(coeffs_and_shifts) < 1e-10] = 0

# remove columns where the coefficients are 0
coeffs_and_shifts = coeffs_and_shifts[:, ~np.all(coeffs_and_shifts == 0, axis=0)]

# sort columns in ascending order according to abs(shift)
coeffs_and_shifts = coeffs_and_shifts[:, np.argsort(np.abs(coeffs_and_shifts)[1])]
```